include/envoy/http/async_client.h

```diff
@@ -38,7 +38,40 @@ class AsyncClient {
   };
 
   /**
-   * An in-flight HTTP request
+   * Notifies caller of async HTTP stream status.
+   */
+  class StreamCallbacks {
+  public:
+    virtual ~StreamCallbacks() {}
+
+    /**
+     * Called when the async HTTP stream receives headers.
+     * @param headers the headers received
+     * @param end_stream whether the response is header only
+     */
+    virtual void onHeaders(HeaderMapPtr&& headers, bool end_stream) PURE;
+
+    /**
+     * Called when the async HTTP stream receives data.
+     * @param data the data received
+     * @param end_stream whether the data is the last data frame
+     */
+    virtual void onData(Buffer::Instance& data, bool end_stream) PURE;
```

Comment:
Hmm... the more I look at this the less convinced I am. The StreamXXXFilter methods are more like the ones I would expect a client to implement and the StreamXXXFilterCallbacks look more like the write interface.

@mattklein123 can probably shed some like on the relationships and why things are implemented this way.

For pure client use-cases I would expect an API more like the one @lizan wrote.